---
title: "Predicting tumor subgroup in pseudobulk and single cells"
author: "Steven Foltz"
date: "2023"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
---

This notebook uses a medulloblastoma subgroup prediction model trained on bulk data to predict the subgroup of pseudobulk samples and individual cells from single-cell RNA-seq experiments.
Pseudobulk samples were generated from single-cell experiments by averaging gene expression across all cells on a per gene level.
Prediction models were trained using bulk gene expression data (bulk array and RNA-seq) in `analysis_notebooks/baseline_models.Rmd`.
Here, we use the trained kTSP and random forest (RF) models to predict single cell subgroups from scRNA-seq data generated by [Hovestadt, et. al](https://www.nature.com/articles/s41586-019-1434-6).

# Setup
```{r setup}

set.seed(8222)

# Load libraries 
source(here::here("utils/color_schemes.R"))
library(patchwork)

# Directories
processed_data_dir <- here::here("processed_data")
plots_dir <- here::here("plots")
plots_data_dir <- here::here(plots_dir, "data")

# Input files
pseudobulk_metadata_filepath <- here::here(processed_data_dir, "pseudobulk_metadata.tsv")
single_cell_plot_data_filepath <- here::here(plots_data_dir, "single_cell_test_predictions.tsv")
pseudobulk_plot_data_filepath <- here::here(plots_data_dir, "pseudobulk_test_predictions.tsv")

# Output plots
pseudobulk_barplot_plot_filepath <- here::here(plots_dir, "single_cell_pseudobulk_score_barplot.pdf")
pseudobulk_points_plot_filepath <- here::here(plots_dir, "single_cell_pseudobulk_score_points.pdf")
percentage_subgroups_plot_filepath <- here::here(plots_dir, "single_cell_percentage_subgroups.pdf")
subgroup_umap_plot_filepath <- here::here(plots_dir, "single_cell_subgroup_umap.pdf")
rf_prob_umap_plot_filepath <- here::here(plots_dir, "single_cell_rf_prob_umap.pdf")
confidence_plot_filepath <- here::here(plots_dir, "single_cell_confidence.pdf")
cluster_umap_plot_filepath <- here::here(plots_dir, "single_cell_cluster_umap.pdf")
cell_prob_plot_filepath <- here::here(plots_dir, "single_cell_rf_prob_barplot.pdf")

# Read in data
pseudobulk_metadata_df <- readr::read_tsv(pseudobulk_metadata_filepath,
                                          show_col_types = FALSE) |>
  dplyr::mutate(sample_accession = title)

pseudobulk_plot_df <- readr::read_tsv(pseudobulk_plot_data_filepath,
                                      show_col_types = FALSE)

single_cell_plot_df <- readr::read_tsv(single_cell_plot_data_filepath,
                                       show_col_types = FALSE)

mb_subgroups <- c("G3", "G4", "SHH", "WNT")

```


# Plot pseudobulk subgroup predictions

Prepare the data for plotting.

```{r}
# Calculate the % of total score for each subgroup for the kTSP model
# This will allow us to use 0-100 for the x-axis
ktsp_pseudobulk_plot_df <- pseudobulk_plot_df |>
  dplyr::filter(model_type == "kTSP (w)") |>
  dplyr::rowwise() |> 
  dplyr::mutate(total_score = sum(G3, G4, SHH, WNT)) |>
  dplyr::mutate_at(dplyr::vars(G3, G4, SHH, WNT), ~ (.x/total_score) * 100) |>
  dplyr::select(-total_score) |>
  dplyr::ungroup() |>
  # Long format for plotting
  tidyr::pivot_longer(cols = all_of(mb_subgroups),
                      names_to = "predicted_subgroup")
  
# The "scores" for RF models are probabilities, given the way we train RF 
# models, so we only need to get this into long format for plotting
rf_pseudobulk_plot_df <- pseudobulk_plot_df |>
  dplyr::filter(model_type == "RF (w)") |>
  tidyr::pivot_longer(cols = all_of(mb_subgroups),
                      names_to = "predicted_subgroup")
```

Create a stacked barplot figure.

```{r}
# Create stacked barplot for RF model
rf_barplot_object <- rf_pseudobulk_plot_df |>
  ggplot2::ggplot(ggplot2::aes(y = sample_accession,
                               x = value,
                               fill = predicted_subgroup)) +
  ggplot2::geom_col() +
  ggplot2::scale_x_continuous(breaks = scales::pretty_breaks()) +
  ggplot2::scale_fill_manual(values = subgroup_colors) +
  ggplot2::facet_grid(rows = ggplot2::vars(subgroup),
                      scales= "free_y",
                      space = "free_y") +
  ggplot2::labs(x = "Probability",
                y = NULL,
                fill = "Predicted Subgroup",
                title = "Random Forest (Weighted)") +
  ggplot2::theme_bw() +
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5,
                                                    size = 12))

# Stacked barplot for kTSP model
ktsp_barplot_object <- ktsp_pseudobulk_plot_df |>
  ggplot2::ggplot(ggplot2::aes(y = sample_accession,
                               x = value,
                               fill = predicted_subgroup)) +
  ggplot2::geom_col() +
  ggplot2::scale_x_continuous(breaks = scales::pretty_breaks()) +
  ggplot2::scale_fill_manual(values = subgroup_colors) +
  ggplot2::facet_grid(rows = ggplot2::vars(subgroup),
                      scales= "free_y",
                      space = "free_y") +
  ggplot2::labs(x = "Percentage of Total Score",
                y = NULL,
                fill = "Predicted Subgroup",
                title = "k Top Scoring Pairs (Weighted)") +
  ggplot2::theme_bw() +
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5,
                                                    size = 12))

# Plot side-by-side with a single legend on the right
pseudobulk_barplot_object <- ktsp_barplot_object + rf_barplot_object + 
  plot_layout(guides = "collect")

ggplot2::ggsave(filename = pseudobulk_barplot_plot_filepath,
                plot = pseudobulk_barplot_object,
                width = 7.5,
                height = 6)

pseudobulk_barplot_object
```

Plot using points, rather than a stacked bar plot.

```{r}
# Points plot for RF model
rf_points_object <- rf_pseudobulk_plot_df |>
  ggplot2::ggplot(ggplot2::aes(y = sample_accession,
                               x = value,
                               color = predicted_subgroup,
                               shape = predicted_subgroup)) +
  ggplot2::geom_point(alpha = 0.75) +
  ggplot2::scale_x_continuous(breaks = scales::pretty_breaks()) +
  ggplot2::scale_color_manual(values = subgroup_colors) +
  ggplot2::facet_grid(rows = ggplot2::vars(subgroup),
                      scales= "free_y",
                      space = "free_y") +
  ggplot2::labs(x = "Probability",
                y = NULL,
                fill = "Predicted Subgroup",
                title = "Random Forest (Weighted)") +
  ggplot2::theme_bw() +
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5,
                                                    size = 12))

# Points plot for kTSP model
ktsp_points_object <- ktsp_pseudobulk_plot_df |>
 ggplot2::ggplot(ggplot2::aes(y = sample_accession,
                               x = value,
                               color = predicted_subgroup,
                               shape = predicted_subgroup)) +
  ggplot2::geom_point(alpha = 0.75) +
  ggplot2::scale_x_continuous(breaks = scales::pretty_breaks()) +
  ggplot2::scale_color_manual(values = subgroup_colors) +
  ggplot2::facet_grid(rows = ggplot2::vars(subgroup),
                      scales= "free_y",
                      space = "free_y") +
  ggplot2::labs(x = "Percentage of Total Score",
                y = NULL,
                fill = "Predicted Subgroup",
                title = "k Top Scoring Pairs (Weighted)") +
  ggplot2::theme_bw() +
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5,
                                                    size = 12))

# Plot side-by-side using a single legend to the right
pseudobulk_points_object <- ktsp_points_object + rf_points_object + 
  plot_layout(guides = "collect")


ggplot2::ggsave(filename = pseudobulk_points_plot_filepath,
                plot = pseudobulk_points_object,
                width = 7.5,
                height = 6)

pseudobulk_points_object
```


### Mislabeled samples

Pull out the mislabeled pseudobulk samples.

```{r}

pseudobulk_plot_df |>
  dplyr::rowwise() |>
  dplyr::mutate(
    predicted_subgroup = c("G3", 
                           "G4", 
                           "SHH", 
                           "WNT")[which.max(c(G3, G4, SHH, WNT))]
  ) |>
  dplyr::ungroup() |>
  dplyr::filter(subgroup != predicted_subgroup) |>
  dplyr::select(Sample = sample_accession,
                `Actual subgroup` = subgroup,
                `Predicted subgroup` = predicted_subgroup,
                `Model` = model_type) |>
   knitr::kable()

```

### Comparison of pseudobulk kTSP vs. RF

- RF models tend to assign some likelihood to every subgroup (median of 4) while kTSP predictions are more frugal (median of 2)
- kTSP narrowly mixed up two G3/G4 cases (BCH825, SJ625) while RF missed only one (SJ625)

# Plot single cell subgroup predictions

```{r}

percentage_subgroups_plot_object <- single_cell_plot_df |>
  dplyr::count(sample_accession, subgroup, predicted_labels, model_type) |>
  ggplot2::ggplot(ggplot2::aes(y = sample_accession,
                               x = n,
                               fill = predicted_labels)) +
  ggplot2::geom_col(position = ggplot2::position_fill()) +
  ggplot2::scale_x_continuous(labels = scales::percent) + # percentage makes sense here
  ggplot2::facet_grid(rows = ggplot2::vars(subgroup),
                      cols = ggplot2::vars(model_type),
                      space = "free_y",
                      scales = "free_y") +
  ggplot2::scale_fill_manual(values = subgroup_colors) +
  ggplot2::labs(y = NULL,
                x = "Single-cell Percentage",
                fill = "Predicted Subgroup") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom")

ggplot2::ggsave(filename = percentage_subgroups_plot_filepath,
                plot = percentage_subgroups_plot_object,
                height = 10,
                width = 7.5)

percentage_subgroups_plot_object

```

### Comparison of single cell kTSP vs. RF

- Most samples show the same tumor subgroup predicted for most of their cells
- kTSP shows greater mixture of predicted subgroups in G3, G4, and SHH compared to RF, less mixture in WNT
- Three samples ("BCH825", "Med2312FH", "SJ625") show largest degree of mixture of predicted subgroups (G3/G4)

# Plot confidence of subgroup predictions

### Prep plotting data

Three samples in particular are of interest given the mixture of subgroups predicted ("BCH825", "Med2312FH", "SJ625").

```{r}

mixed_samples <- sort(c("BCH825", "Med2312FH", "SJ625"))

mixed_samples_with_subgroup <- pseudobulk_metadata_df |>
  dplyr::filter(sample_accession %in% mixed_samples) |>
  dplyr::mutate(sample_with_subgroup = stringr::str_c(title, " (", subgroup, ")")) |>
  dplyr::select(sample_accession, subgroup, sample_with_subgroup)

```

### Plot UMAPs with subgroup prediction

```{r}

subgroup_umap_plot_object <- single_cell_plot_df |>
  dplyr::filter(sample_accession %in% mixed_samples) |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  ggplot2::ggplot(ggplot2::aes(x = UMAP_1,
                               y = UMAP_2,
                               color = predicted_labels)) +
  ggplot2::geom_point(shape = 16,
                      alpha = 0.5) +
  ggplot2::scale_color_manual(values = subgroup_colors) +
  ggplot2::scale_alpha_continuous(range = c(0, 1)) +
  ggplot2::facet_grid(rows = ggplot2::vars(sample_with_subgroup),
                      cols = ggplot2::vars(model_type),
                      scales = "free") +
  ggplot2::labs(x = "UMAP 1",
                y = "UMAP 2",
                color = "Predicted\nSubgroup") +
  ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(alpha = 1))) +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom",
                 legend.direction = "horizontal")

ggplot2::ggsave(filename = subgroup_umap_plot_filepath,
                plot = subgroup_umap_plot_object,
                width = 7.5,
                height = 6)

subgroup_umap_plot_object

```

### Plot UMAPs with RF probabilities

```{r}
rf_prob_umap_object <- single_cell_plot_df |>
  dplyr::filter(sample_accession %in% mixed_samples,
                model_type == "RF (w)") |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  dplyr::select(sample_with_subgroup, 
                G3,
                G4,
                UMAP_1,
                UMAP_2) |>
  tidyr::pivot_longer(cols = c(G3, G4),
                      names_to = "predicted_subgroup") |>
  ggplot2::ggplot(ggplot2::aes(x = UMAP_1,
                               y = UMAP_2,
                               color = value)) +
  ggplot2::geom_point(shape = 16,
                      alpha = 0.8) +
  ggplot2::facet_grid(rows = ggplot2::vars(sample_with_subgroup),
                      cols = ggplot2::vars(predicted_subgroup)) +
  ggplot2::scale_color_viridis_c() +
  ggplot2::theme_bw() +
  ggplot2::labs(x = "UMAP 1",
                y = "UMAP 2",
                color = "Probability",
                title = "Random Forest (Weighted)") +
  ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5))

ggplot2::ggsave(filename = rf_prob_umap_plot_filepath,
                plot = rf_prob_umap_object,
                width = 7.5,
                height = 6)

rf_prob_umap_object

```


### Plot UMAPs with assigned clusters

```{r}

cluster_umap_plot_object <- single_cell_plot_df |>
  dplyr::filter(sample_accession %in% mixed_samples) |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  ggplot2::ggplot(ggplot2::aes(x = UMAP_1,
                               y = UMAP_2,
                               color = factor(cluster))) +
  ggplot2::geom_point(shape = 16,
                      alpha = 0.5) +
  ggplot2::facet_grid(rows = ggplot2::vars(sample_with_subgroup),
                      cols = ggplot2::vars(model_type),
                      scales = "free") +
  ggplot2::scale_color_brewer(palette = "Set1") +
  ggplot2::labs(x = "UMAP 1",
                y = "UMAP 2",
                color = "Cluster") +
  ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(alpha = 1))) +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom",
                 legend.direction = "horizontal")

ggplot2::ggsave(filename = cluster_umap_plot_filepath,
                plot = cluster_umap_plot_object,
                width = 7.5,
                height = 6)

cluster_umap_plot_object

```

### Plot confidence of individual cell predictions

```{r}

typical_confidence_df <- single_cell_plot_df |>
  dplyr::filter(!is.na(confidence),
                model_type == "kTSP (w)") |>
  dplyr::group_by(sample_accession,
                  subgroup,
                  predicted_labels,
                  model_type) |>
  dplyr::summarize(median_confidence = median(confidence),
                   .groups = "drop") |>
  tidyr::pivot_wider(names_from = predicted_labels,
                     values_from = median_confidence) |>
  dplyr::rowwise() |>
  dplyr::mutate(subgroup_conf = c(G3, G4, SHH, WNT)[which(mb_subgroups == subgroup)]) |>
  dplyr::ungroup() |>
  dplyr::group_by(model_type) |>
  dplyr::summarize(median_subgroup_conf = median(subgroup_conf))

cell_counts_df <- single_cell_plot_df |>
  dplyr::filter(sample_accession %in% mixed_samples,
                !is.na(confidence),
                model_type == "kTSP (w)") |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  dplyr::count(sample_with_subgroup,
               model_type,
               predicted_labels)

confidence_plot_object <- single_cell_plot_df |>
  dplyr::filter(sample_accession %in% mixed_samples,
                !is.na(confidence),
                model_type == "kTSP (w)") |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  ggplot2::ggplot(ggplot2::aes(x = predicted_labels,
                               y = confidence,
                               color = predicted_labels)) +
  ggplot2::geom_hline(data = typical_confidence_df,
                      mapping = ggplot2::aes(yintercept = median_subgroup_conf),
                      lty = 2) +
  ggplot2::geom_violin(fill = NA,
                       draw_quantiles = 0.5,
                       position = ggplot2::position_dodge(width = 0.75), scale = "width",
                       show.legend = FALSE) +
  ggplot2::geom_point(shape = 16,
                      alpha = 0.5,
                      position = ggplot2::position_jitterdodge(jitter.height = 0,
                                                               jitter.width = 0.25)) +
  ggplot2::geom_text(data = cell_counts_df,
                     ggplot2::aes(label = n),
                     y = 0,
                     vjust = 0,
                     show.legend = FALSE) +
  ggplot2::facet_wrap(ggplot2::vars(sample_with_subgroup)) +
  ggplot2::scale_color_manual(values = subgroup_colors) +
  ggplot2::scale_y_continuous(limits = c(0, NA)) +
  ggplot2::labs(x = "Predicted Subgroup",
                y = "Confidence",
                color = "Predicted\nSubgroup") +
  ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(alpha = 1))) +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom",
                 legend.direction = "horizontal")

ggplot2::ggsave(filename = confidence_plot_filepath,
                plot = confidence_plot_object,
                width = 7.5, height = 6)

confidence_plot_object

```

- Cell subgroup predictions are more confident when subgroup predictions are consistent across entire sample.
- The relative confidence levels of each cell's prediction show that predictions in samples with one dominant subgroup are more confident than in samples with mixed subgroup predictions (BCH825, Med2312FH, SJ625).
- While some cells from mixed subgroup samples show high confidence as G3 or G4, most show that cells receive signals from both G3 and G4 rules, meaning they have an intermediate phenotype, rather than being a mixture of true G3 and true G4 cells.

### Individual cell probabilities for RF model

```{r}
# Subset to data needed to plot individual cell probabilities for RF model
individual_cells_prob_df <- single_cell_plot_df |>
  dplyr::filter(sample_accession %in% mixed_samples,
                model_type == "RF (w)") |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  dplyr::select(sample_accession,
                sample_with_subgroup,
                cell_index,
                predicted_labels,
                all_of(mb_subgroups)) |>
  dplyr::mutate(cell_index = as.character(cell_index)) |>
  tidyr::pivot_longer(cols = all_of(mb_subgroups),
                      names_to = "predicted_subgroup")

# Use the sample with subgroup column for display
sample_plot_titles <- unique(individual_cells_prob_df$sample_with_subgroup)

# Make a stacked barplot of probabilities for each sample
cells_prob_plots_list <- purrr::map(mixed_samples,
                                    \(sample_id) individual_cells_prob_df |> 
                                      dplyr::filter(sample_accession == sample_id,
                                                    predicted_labels %in% c("G3", "G4")) |>
                                      ggplot2::ggplot(
                                        ggplot2::aes(y = cell_index,
                                                     x = value,
                                                     fill = predicted_subgroup)) +
                                      ggplot2::geom_col(position = ggplot2::position_fill()) +
                                      ggplot2::scale_x_continuous(breaks = scales::pretty_breaks()) +
                                      ggplot2::scale_fill_manual(values = subgroup_colors) +
                                      ggplot2::facet_grid(rows = ggplot2::vars(predicted_labels),
                                                          scales = "free_y",
                                                          space = "free_y") +
                                      ggplot2::labs(fill = "Predicted Subgroup",
                                                    x = "Probability",
                                                    title = sample_plot_titles[stringr::str_detect(sample_plot_titles, sample_id)]) +
                                      ggplot2::theme_bw() +
                                      ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5,
                                                                                        size = 12),
                                                     axis.title.y = ggplot2::element_blank(),
                                                     axis.text.y = ggplot2::element_blank(),
                                                     axis.ticks.y = ggplot2::element_blank())
)

# Create a 3 column plot with a single legend to the right of the panels
cell_prob_object <- wrap_plots(cells_prob_plots_list, ncol = 3)  + 
  plot_layout(guides = "collect")

ggplot2::ggsave(filename = cell_prob_plot_filepath,
                plot = cell_prob_object,
                width = 7.5,
                height = 6)

cell_prob_object
```

### G3 and G4 prediction confidence exists on a G3-G4 spectrum

```{r}

joint_G3_G4_df <- single_cell_plot_df |>
  dplyr::filter(sample_accession %in% mixed_samples,
                !is.na(confidence)) |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  dplyr::filter(predicted_labels %in% c("G3", "G4"))

# what percentage of cells have no contribution from SHH or WNT (G3 and G4 only)
diagonal_pct_list <- joint_G3_G4_df |>
  split(joint_G3_G4_df$model_type) |>
  purrr::map(\(x) mean(x$SHH == 0 & x$WNT == 0)*100)

joint_G3_G4_confidence_plot_object <- joint_G3_G4_df |>
  ggplot2::ggplot(ggplot2::aes(x = G3/total,
                               y = G4/total,
                               color = predicted_labels)) +
  ggplot2::geom_segment(x = 0, xend = 1,
                        y = 1, yend = 0,
                        color = "grey",
                        lty = 1) +
  ggplot2::geom_point(shape = 16,
                      alpha = 0.5) +
  ggplot2::geom_density_2d(show.legend = FALSE) +
  ggplot2::facet_grid(rows = ggplot2::vars(sample_with_subgroup),
                      cols = ggplot2::vars(model_type),
                      scales = "free") +
  ggplot2::scale_color_manual(values = subgroup_colors) +
  ggplot2::scale_x_continuous(limits = c(0,1)) +
  ggplot2::scale_y_continuous(limits = c(0,1)) +
  ggplot2::labs(x = "G3 Confidence",
                y = "G4 Confidence",
                color = "Predicted Subgroup") +
  ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(alpha = 1))) +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom",
                 legend.direction = "horizontal")

joint_G3_G4_confidence_plot_object

```

- Prediction confidence of G3 and G4 cells forms a spectrum between poles at 100% G3 and 100% G4.
- For kTSP (w), some cells have some score contributed from other subgroups, but most (`r round(diagonal_pct_list[["kTSP (w)"]], 1)`%) are on the diagonal.

# Session info
```{r session_info}
sessionInfo()
```
