---
title: "Train and test models to predict MB subgroups using a targeted gene set"
author: "Steven Foltz"
date: "2023"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
params:
  create_models: FALSE # train new models (if FALSE, reads existing models from file)
  overwrite: FALSE # if create_models is also TRUE, overwrite existing models file
  seed: 44 # set initial seed for run_many_models() and before stochastic plots
  n_repeats: 10 # set number of times to repeat in run_many_models()
  n_cores: 3 # set number of cores to use in run_many_models()
---

# Rationale

# Setup

```{r setup}

# get input parameters
create_models <- params$create_models
overwrite <- params$overwrite

# Source code and libraries
source(here::here("utils/color_schemes.R"))
source(here::here("utils/convert_gene_names.R"))
source(here::here("utils/modeling.R"))
library(patchwork)

# Define directories and input/output file paths
processed_data_dir <- here::here("processed_data")
models_dir <- here::here("models")
plots_dir <- here::here("plots")
plots_data_dir <- file.path(plots_dir, "data")

bulk_genex_filepath <- file.path(processed_data_dir, "bulk_genex.tsv")
bulk_metadata_filepath <- file.path(processed_data_dir, "bulk_metadata.tsv")

baseline_filepath <- file.path(models_dir, "baseline.rds")
targeted_filepath <- file.path(models_dir, "targeted_gene_set.rds")
random_filepath <- file.path(models_dir, "random_gene_sets.rds")

baseline_plot_df_filepath <- file.path(plots_data_dir, "baseline_plots_df.tsv")
targeted_plot_df_filepath <- file.path(plots_data_dir, "targeted_plot_df.tsv")
random_plot_df_filepath <- file.path(plots_data_dir, "random_plot_df.tsv")
targeted_delta_kappa_plot_df_filepath <- file.path(plots_data_dir, "targeted_delta_kappa_plot_df.tsv")

# check that existing targeted gene set model file will not be overwritten if overwrite is FALSE
if (create_models &
    (file.exists(targeted_filepath)) &
    !overwrite) {
  
  stop("Targeted gene set model output file already exists and overwrite is set to FALSE in analysis_notebooks/targeted_gene_set.Rmd.")
  
}

# check that existing random gene sets model file will not be overwritten if overwrite is FALSE
if (create_models &
    (file.exists(random_filepath)) &
    !overwrite) {
  
  stop("Random gene sets model output file already exists and overwrite is set to FALSE in analysis_notebooks/targeted_gene_set.Rmd.")
  
}

# check that targeted file exist if create_models is FALSE
if (!create_models & !file.exists(targeted_filepath)) {
  
  stop("Targeted model output file does not exist and create_models is set to FALSE in analysis_notebooks/targeted_gene_set.Rmd.")
  
}

# check that random file exist if create_models is FALSE
if (!create_models & !file.exists(random_filepath)) {
  
  stop("Random model output file does not exist and create_models is set to FALSE in analysis_notebooks/targeted_gene_set.Rmd.")
  
}

# set subgroups analyzed in this notebook (canonical MB subgroups)
mb_subgroups <- c("G3", "G4", "SHH", "WNT")

# Read in essential data
bulk_genex_df <- readr::read_tsv(bulk_genex_filepath,
                                 show_col_types = FALSE) |>
  tibble::column_to_rownames(var = "gene")

bulk_metadata_df <- readr::read_tsv(bulk_metadata_filepath,
                                    show_col_types = FALSE) |>
  dplyr::filter(sample_accession %in% names(bulk_genex_df),
                subgroup %in% mb_subgroups)

bulk_genex_df <- bulk_genex_df |>
  dplyr::select(dplyr::all_of(bulk_metadata_df$sample_accession))

check_input_files(genex_df = bulk_genex_df,
                  metadata_df = bulk_metadata_df)

# targeted genes
targeted_genes_filepath <- file.path(processed_data_dir,
                                     "NS_IO_360_v1.0_Genes.tsv")
targeted_genes_df <- readr::read_tsv(file = targeted_genes_filepath,
                                     show_col_types = FALSE) |>
  convert_gene_names(gene_column_before = "Gene",
                     gene_column_after = "gene",
                     map_from = "SYMBOL",
                     map_to = "ENSEMBL")

# reduce bulk genex df to overlap with targeted genes
bulk_genex_df_targeted <- bulk_genex_df[row.names(bulk_genex_df) %in% targeted_genes_df$gene, ]

n_targeted_genes_overlap <- nrow(bulk_genex_df_targeted)

# generate list of 10 random gene sets
random_gene_set_list <- purrr::map(1:10,
                                   \(x) sample(row.names(bulk_genex_df),
                                               size = n_targeted_genes_overlap,
                                               replace = FALSE))

# select random genes from bulk genex df as a list
random_gene_set_bulk_genex_df_list <- random_gene_set_list |>
  purrr::map(\(x) bulk_genex_df[row.names(bulk_genex_df) %in% x, ])

```

# Train and test MB subgroup models using targeted gene set

### Can we use MM2S with targeted gene set?


```{r}

targeted_genes_df_ENTREZID <- targeted_genes_df |>
  convert_gene_names(gene_column_before = "gene",
                     gene_column_after = "gene",
                     map_from = "ENSEMBL",
                     map_to = "ENTREZID")

n_mm2s_common_pathways <- purrr::map(MM2S::HumanGMT$genesets,
                                     \(x) sum(x %in% row.names(targeted_genes_df_ENTREZID))) |>
  purrr::map_lgl(\(x) x >= 20 & x <= 100) |>
  sum()

ifelse(n_mm2s_common_pathways >= 24,
       "Yes!", "Unfortunately, no.")

```

MM2S uses a pathway-based approach, and those pathways are pre-defined in `MM2S::HumanGMT$genesets`.
To return a GSVA score for a particular pathway, MM2S requires that between 20 and 100 genes overlap between the test data and the set of genes in that pathway.
MM2S later requires that there are at least 24 such pathways in the data.
However, with our targeted gene set, only `r n_mm2s_common_pathways` pathways match the `20 <= overlapping genes <= 100` criteria.
MM2S fails to run because an `NA` value is introduced to `FeatureSelection` and `NorthcottFeatures` inside `MM2S::MM2S.human`, resulting in invalid column selection in `HumanTestGSVA <- HumanTestGSVA[, NorthcottFeatures, drop = FALSE]`.

### Create kTSP, RF, and LASSO models

The following code creates weighted kTSP and RF models only.
LASSO models use default settings.
Supplying the same `initial_seed` to each instance of `run_many_models()` ensures the training/testing sets of each repeat are the same for each model.

```{r}

model_types <- c("ktsp", "rf", "lasso")

if (create_models) {
  
  # uses the bulk_genex_df reduced to targeted gene set
  targeted_list <- run_many_models(genex_df = bulk_genex_df_targeted,
                                   metadata_df = bulk_metadata_df,
                                   labels = mb_subgroups,
                                   model_types = model_types,
                                   initial_seed = params$seed,
                                   n_repeats = params$n_repeats,
                                   n_cores = params$n_cores,
                                   ktsp_featureNo = 1000,
                                   ktsp_n_rules_min = 5,
                                   ktsp_n_rules_max = 50,
                                   ktsp_weighted = TRUE,
                                   rf_num.trees = 500,
                                   rf_genes_altogether = 50,
                                   rf_genes_one_vs_rest = 50,
                                   rf_gene_repetition = 1,
                                   rf_rules_altogether = 50,
                                   rf_rules_one_vs_rest = 50,
                                   rf_weighted = TRUE)
  
  # write models to file
  readr::write_rds(x = targeted_list,
                   file = targeted_filepath)
  
  # uses the bulk_genex_df reduced to random gene sets
  random_list <- random_gene_set_bulk_genex_df_list |>
    purrr::map(\(x) run_many_models(genex_df = x,
                                    metadata_df = bulk_metadata_df,
                                    labels = mb_subgroups,
                                    model_types = model_types,
                                    initial_seed = params$seed,
                                    n_repeats = params$n_repeats,
                                    n_cores = params$n_cores,
                                    ktsp_featureNo = 1000,
                                    ktsp_n_rules_min = 5,
                                    ktsp_n_rules_max = 50,
                                    ktsp_weighted = TRUE,
                                    rf_num.trees = 500,
                                    rf_genes_altogether = 50,
                                    rf_genes_one_vs_rest = 50,
                                    rf_gene_repetition = 1,
                                    rf_rules_altogether = 50,
                                    rf_rules_one_vs_rest = 50,
                                    rf_weighted = TRUE))
               
  # write models to file
  readr::write_rds(x = random_list,
                   file = random_filepath)
  
} else { # if create_models = FALSE and models already exist
  
  targeted_list <- readr::read_rds(targeted_filepath)
  random_list <- readr::read_rds(random_filepath)
  
}

```

# Visualize Kappa and accuracy results

### Gather targeted data

```{r}

targeted_plot_df <- targeted_list |>
  purrr::map(return_model_metrics,
             model_types = model_types,
             metadata_df = bulk_metadata_df,
             labels = mb_subgroups,
             platforms = c("Array", "RNA-seq")) |>
  purrr::list_rbind(names_to = "repeat_index") |>
  dplyr::filter(metric == "Kappa") |>
  dplyr::mutate(model_type = dplyr::case_match(model_type,
                                               "ktsp" ~ "kTSP (w)",
                                               "rf" ~ "RF (w)",
                                               "lasso" ~ "LASSO"))

readr::write_tsv(x = targeted_plot_df,
                 file = targeted_plot_df_filepath)
```

### Gather random data

```{r}

random_plot_df <- random_list |>
  purrr::map(\(x) x |>
               purrr::map(return_model_metrics,
                          model_types = model_types,
                          metadata_df = bulk_metadata_df,
                          labels = mb_subgroups,
                          platforms = c("Array", "RNA-seq")) |>
               purrr::list_rbind(names_to = "repeat_index") |>
               dplyr::filter(metric == "Kappa") |>
               dplyr::mutate(model_type = dplyr::case_match(model_type,
                                                            "ktsp" ~ "kTSP (w)",
                                                            "rf" ~ "RF (w)",
                                                            "lasso" ~ "LASSO"))) |>
  purrr::list_rbind(names_to = "random_index")

readr::write_tsv(x = random_plot_df,
                 file = random_plot_df_filepath)

```

# Find delta kappa values

```{r}

baseline_plot_df <- readr::read_tsv(file = baseline_plot_df_filepath,
                                    show_col_types = FALSE) |>
  dplyr::filter(metric == "Kappa")

targeted_delta_kappa_plot_df <- targeted_plot_df |>
  dplyr::select(repeat_index,
                model_type,
                platform,
                targeted_kappa = value) |>
  dplyr::left_join(baseline_plot_df |>
                     dplyr::select(repeat_index,
                                   model_type,
                                   platform,
                                   baseline_kappa = value),
                   by = c("repeat_index",
                          "model_type",
                          "platform")) |>
  dplyr::mutate(delta_kappa = targeted_kappa - baseline_kappa) |>
  readr::write_tsv(file = targeted_delta_kappa_plot_df_filepath)

set.seed(1)

targeted_delta_kappa_plot_object <- targeted_delta_kappa_plot_df |>
  ggplot2::ggplot(ggplot2::aes(x = model_type,
                               y = delta_kappa,
                               color = platform)) +
  ggplot2::geom_violin(position = ggplot2::position_dodge(width = 0.7),
                       show.legend = FALSE) +
  ggplot2::geom_point(position = ggplot2::position_jitterdodge(jitter.width = 0.1,
                                                               jitter.height = 0,
                                                               dodge.width = 0.7)) +
  ggplot2::scale_color_manual(values = platform_colors) +
  ggplot2::labs(x = "Method",
                y = "Delta Kappa",
                color = "Platform",
                title = "Change in Kappa with Targeted Gene Set") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom",
                 legend.direction = "horizontal")

targeted_delta_kappa_plot_object

```

# Session info

```{r session_info}
sessionInfo()
```
