---
title: "Train and test models to predict MB subgroups using a targeted gene set"
author: "Steven Foltz"
date: "2023"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
params:
  create_models: FALSE # train new models (if FALSE, reads existing models from file)
  overwrite: FALSE # if create_models is also TRUE, overwrite existing models file
  seed: 44 # set initial seed for run_many_models() and before stochastic plots
  n_repeats: 10 # set number of times to repeat in run_many_models()
  n_cores: 3 # set number of cores to use in run_many_models()
---

# Rationale

# Setup

```{r setup}

# get input parameters
create_models <- params$create_models
overwrite <- params$overwrite

# Source code and libraries
source(here::here("utils/color_schemes.R"))
source(here::here("utils/convert_gene_names.R"))
source(here::here("utils/modeling.R"))
library(patchwork)

# Define directories and input/output file paths
processed_data_dir <- here::here("processed_data")
models_dir <- here::here("models")
plots_dir <- here::here("plots")
plots_data_dir <- file.path(plots_dir, "data")

bulk_genex_filepath <- file.path(processed_data_dir, "bulk_genex.tsv")
bulk_metadata_filepath <- file.path(processed_data_dir, "bulk_metadata.tsv")

baseline_filepath <- file.path(models_dir, "baseline.rds")
targeted_filepath <- file.path(models_dir, "targeted_gene_set.rds")
random_filepath <- file.path(models_dir, "random_gene_sets.rds")

# check that existing targeted gene set model file will not be overwritten if overwrite is FALSE
if (create_models &
    (file.exists(targeted_filepath)) &
    !overwrite) {
  
  stop("Targeted gene set model output file already exists and overwrite is set to FALSE in analysis_notebooks/targeted_gene_set.Rmd.")
  
}

# check that existing random gene sets model file will not be overwritten if overwrite is FALSE
if (create_models &
    (file.exists(random_filepath)) &
    !overwrite) {
  
  stop("Random gene sets model output file already exists and overwrite is set to FALSE in analysis_notebooks/targeted_gene_set.Rmd.")
  
}

# check that targeted file exist if create_models is FALSE
if (!create_models & !file.exists(targeted_filepath)) {
  
  stop("Targeted model output file does not exist and create_models is set to FALSE in analysis_notebooks/targeted_gene_set.Rmd.")
  
}

# check that random file exist if create_models is FALSE
if (!create_models & !file.exists(random_filepath)) {
  
  stop("Random model output file does not exist and create_models is set to FALSE in analysis_notebooks/targeted_gene_set.Rmd.")
  
}

# set subgroups analyzed in this notebook (canonical MB subgroups)
mb_subgroups <- c("G3", "G4", "SHH", "WNT")

# Read in essential data
bulk_genex_df <- readr::read_tsv(bulk_genex_filepath,
                                 show_col_types = FALSE) |>
  tibble::column_to_rownames(var = "gene")

bulk_metadata_df <- readr::read_tsv(bulk_metadata_filepath,
                                    show_col_types = FALSE) |>
  dplyr::filter(sample_accession %in% names(bulk_genex_df),
                subgroup %in% mb_subgroups)

bulk_genex_df <- bulk_genex_df |>
  dplyr::select(dplyr::all_of(bulk_metadata_df$sample_accession))

check_input_files(genex_df = bulk_genex_df,
                  metadata_df = bulk_metadata_df)

# targeted genes
targeted_genes_filepath <- file.path(processed_data_dir,
                                     "NS_IO_360_v1.0_Genes.tsv")
targeted_genes_df <- readr::read_tsv(file = targeted_genes_filepath,
                                     show_col_types = FALSE) |>
  convert_gene_names(gene_column_before = "Gene",
                     gene_column_after = "gene",
                     map_from = "SYMBOL",
                     map_to = "ENSEMBL")

# reduce bulk genex df to overlap with targeted genes
bulk_genex_df_targeted <- bulk_genex_df[row.names(bulk_genex_df) %in% targeted_genes_df$gene, ]

n_targeted_genes_overlap <- nrow(bulk_genex_df_targeted)

# generate list of 10 random gene sets
random_gene_set_list <- purrr::map(1:10,
                                   \(x) sample(row.names(bulk_genex_df),
                                               size = n_targeted_genes_overlap,
                                               replace = FALSE))

# select random genes from bulk genex df as a list
random_gene_set_bulk_genex_df_list <- random_gene_set_list |>
  purrr::map(\(x) bulk_genex_df[row.names(bulk_genex_df) %in% x, ])

```

# Train and test MB subgroup models using targeted gene set

### Can we use MM2S with targeted gene set?


### Create kTSP, RF, and LASSO models

The following code creates weighted kTSP and RF models only.
LASSO models use default settings.
Supplying the same `initial_seed` to each instance of `run_many_models()` ensures the training/testing sets of each repeat are the same for each model.

```{r}

model_types <- c("ktsp_weighted",
                 "rf_weighted",
                 "lasso")

if (create_models) {
  
  # uses the bulk_genex_df reduced to targeted gene set
  targeted_list <- run_many_models(genex_df = bulk_genex_df_targeted,
                                   metadata_df = bulk_metadata_df,
                                   labels = mb_subgroups,
                                   model_types = c("ktsp", "rf", "lasso"),
                                   initial_seed = params$seed,
                                   n_repeats = params$n_repeats,
                                   n_cores = params$n_cores,
                                   ktsp_featureNo = 1000,
                                   ktsp_n_rules_min = 5,
                                   ktsp_n_rules_max = 50,
                                   ktsp_weighted = TRUE,
                                   rf_num.trees = 500,
                                   rf_genes_altogether = 50,
                                   rf_genes_one_vs_rest = 50,
                                   rf_gene_repetition = 1,
                                   rf_rules_altogether = 50,
                                   rf_rules_one_vs_rest = 50,
                                   rf_weighted = TRUE)
  
  # write models to file
  readr::write_rds(x = targeted_list,
                   file = targeted_filepath)
  
  # uses the bulk_genex_df reduced to random gene sets
  random_list <- random_gene_set_bulk_genex_df_list |>
    purrr::map(\(x) run_many_models(genex_df = x,
                                    metadata_df = bulk_metadata_df,
                                    labels = mb_subgroups,
                                    model_types = c("ktsp", "rf", "lasso"),
                                    initial_seed = params$seed,
                                    n_repeats = params$n_repeats,
                                    n_cores = params$n_cores,
                                    ktsp_featureNo = 1000,
                                    ktsp_n_rules_min = 5,
                                    ktsp_n_rules_max = 50,
                                    ktsp_weighted = TRUE,
                                    rf_num.trees = 500,
                                    rf_genes_altogether = 50,
                                    rf_genes_one_vs_rest = 50,
                                    rf_gene_repetition = 1,
                                    rf_rules_altogether = 50,
                                    rf_rules_one_vs_rest = 50,
                                    rf_weighted = TRUE))
               
  # write models to file
  readr::write_rds(x = targeted_list,
                   file = targeted_filepath)
  
} else { # if create_models = FALSE and models already exist
  
  targeted_list <- readr::read_rds(targeted_filepath)
  random_list <- readr::read_rds(random_filepath)
  
}

stop()

```

# Visualize Kappa and accuracy results

### Gather data

The variable `baseline_list` is a list of lists.
The first level of lists correspond to different repeats using different training/testing sets.
Within each repeat, there are list elements corresponding to particular model results (e.g., `ktsp_weighted` or `lasso`) as well as metadata about the repeat (e.g., the seed used for to determine training/testing sets).
To create a data frame for plotting, we need to gather modeling metrics (Kappa and Balanced Accuracy) from each model across all the repeats.

```{r}
plot_df <- baseline_list |>
  purrr::map(return_model_metrics,
             model_types = model_types,
             metadata_df = bulk_metadata_df,
             labels = mb_subgroups,
             platforms = c("Array", "RNA-seq")) |>
  purrr::list_rbind(names_to = "repeat_index") |>
  dplyr::mutate(model_type = dplyr::case_match(model_type,
                                               "ktsp_unweighted" ~ "kTSP (unw)",
                                               "ktsp_weighted" ~ "kTSP (w)",
                                               "rf_unweighted" ~ "RF (unw)",
                                               "rf_weighted" ~ "RF (w)",
                                               "mm2s" ~ "MM2S",
                                               "lasso" ~ "LASSO"))
readr::write_tsv(x = plot_df,
                 file = baseline_plot_df_filepath)
```

### Kappa tables

Kappa measures how often our predicted subgroups agree with the true subgroups, accounting for the chance of random agreement (necessary when subgroups are imbalanced).
Kappa is an overall measure that applies to an entire set of predictions across subgroups.
Values of Kappa closer to 1 indicate higher agreement, and values closer to 0 indicate random assignment.

$$ \kappa = \frac{p_o - p_e}{1 - p_e} $$

where $p_o$ is the observed proportion of agreement and $p_e$ is the expected proportion of chance agreement.

The following table summarizes Kappa values stratified by platform (Array or RNA-seq) and model type by showing the median Kappa value.

```{r}
plot_df |>
  dplyr::filter(metric == "Kappa") |>
  dplyr::group_by(platform, model_type) |>
  dplyr::summarize(median_kappa = median(value),
                   .groups = "drop") |>
  tidyr::pivot_wider(names_from = platform,
                     values_from = median_kappa) |>
  knitr::kable(caption = "Median Kappa values (by platform)")
```

- Models perform slightly better on Array test data (mid-.90s) than RNA-seq test data (low-.90s) (except for MM2S).
- kTSP (w) and kTSP (unw) show no difference in performance in Array test data (more Array data in training set), but kTSP (unw) outperformed kTSP (w) in RNA-seq test data (against expectation)
- The reverse is true for RF models, with RF (w) outperforming RF (unw) for both Array and RNA-seq test data
- In weighted models, RF outperformed kTSP in RNA-seq, while the unweighted kTSP outperformed RF in both platforms
- MM2S showed better test performance in RNA-seq data despite the original model being trained on Array


# Session info
```{r session_info}
sessionInfo()
```
