---
title: "Predicting tumor subgroup in single cells"
author: "Steven Foltz"
date: "2023"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
---

This notebook uses a medulloblastoma subgroup prediction model to predict the subgroup of individual cells from single-cell RNA-seq experiments.
Models were trained using bulk gene expression data (bulk array and RNA-seq) in `analysis_notebooks/baseline_models.Rmd`.
Here, we use the trained model to predict single cell subgroups from scRNA-seq data generated by [Hovestadt, et. al](https://www.nature.com/articles/s41586-019-1434-6).


# Setup
```{r setup}

# Load libraries 
source(here::here("utils/modeling.R"))
source(here::here("utils/single-cell.R"))

# Input files

processed_data_dir <- here::here("processed_data")
models_dir <- here::here("models")
pseudobulk_sce_dir <- file.path(processed_data_dir, "pseudobulk_sce")

pseudobulk_metadata_filepath <- file.path(processed_data_dir, "pseudobulk_metadata.tsv")

gene_map_filepath <- file.path(processed_data_dir, "gene_map.tsv")

# baseline models are generated prior to this in the notebook analysis_notebooks/baseline_models.Rmd
# there is one baseline model designated as the official model that is used to test single cells here
# "baseline" means that we used all genes and all available training/test data in modeling
baseline_models_filepath <- file.path(models_dir, "baseline.rds")

# Output files
plots_dir <- here::here("plots")
relative_confidence_plot_filepath <- file.path(plots, "single_cell_relative_confidence.pdf")
combined_umap_plot_filepath <- file.path(plots, "single_cell_combined_umap.pdf")

# Read in data
pseudobulk_metadata_df <- readr::read_tsv(pseudobulk_metadata_filepath,
                                          show_col_types = FALSE) |>
  dplyr::mutate(sample_accession = title) |>
  dplyr::mutate(platform = "RNA-seq")

gene_map_df <- readr::read_tsv(gene_map_filepath,
                               show_col_types = FALSE)

baseline_models <- readr::read_rds(baseline_models_filepath)
ktsp_classifier <- baseline_models[[1]]$ktsp$classifier #separate issue: pick the designated model

```

# Predict subgroup of single cells

```{r message = FALSE}

test_object_list <- lapply(pseudobulk_metadata_df$title,
                           function(x) test_single_cells(sample_acc = x,
                                                         sce_filepath = file.path(pseudobulk_sce_dir, stringr::str_c(x, "_sce.rds")),
                                                         metadata_df = pseudobulk_metadata_df,
                                                         gene_map_df = gene_map_df,
                                                         ktsp_classifier = ktsp_classifier)) |>
  purrr::set_names(pseudobulk_metadata_df$title)

```

### Prep plotting data

Several samples in particular are of interest given the mixture of subgroups predicted ("BCH825", "Med2312FH", "SJ625").
Four other samples can serve as representative samples from each subgroup ("BCH1205", "BCH1031", "MUV41", "BCH807").

```{r}

sample_titles <- c("BCH825", "Med2312FH", "SJ625", "BCH1205", "BCH1031", "MUV41", "BCH807")
sample_titles_with_subgroup <- pseudobulk_metadata_df %>%
  filter(title %in% sample_titles) %>%
  mutate(title_with_subgroup = stringr::str_c(title, " (", subgroup, ")")) %>%
  pull(title_with_subgroup)

umap_labels_combined_df <- NULL
for(i in 1:length(sample_titles)) {
  
  sample_title <- sample_titles[i]
  sample_title_with_subgroup <- sample_titles_with_subgroup[i]
  
  sce_filepath <- file.path(pseudobulk_sce_dir, stringr::str_c(sample_title, "_sce.rds"))
  sce_object <- readr::read_rds(sce_filepath)
  umap_df <- data.frame(sample = sample_title,
                        sample_with_subgroup = sample_title_with_subgroup,
                        sce_object@int_colData$reducedDims$UMAP,
                        cluster = sce_object@colData$louvain_10)
  names(umap_df) <- c("sample", "UMAP_1", "UMAP_2", "cluster")
  
  labels_df <- test_object_list[[sample_title]][["model_output"]] |>
    tibble::as_tibble() |>
    dplyr::rowwise() |>
    dplyr::mutate(max = max(G3, G4, SHH, WNT),
                  total = sum(G3, G4, SHH, WNT),
                  other = total - max,
                  confidence = max/total) |>
    dplyr::ungroup() |>
    dplyr::mutate(confidence = max/total,
                  relative_confidence = max/(total - max))
  
  umap_labels_combined_df <- dplyr::bind_rows(umap_labels_combined_df,
                                              tibble::as_tibble(dplyr::bind_cols(umap_df, labels_df)))
  
}

```

# Plot results

### How confident are individual cell predictions?
More confident when subgroup predictions are consistent across entire sample.
```{r fig.height = 5, fig.width = 7.5}

relative_confidence_plot_object <- umap_labels_combined_df |>
  ggplot2::ggplot(ggplot2::aes(x = cluster,
                               y = relative_confidence,
                               color = max_score)) +
  ggplot2::geom_hline(yintercept = c(1,2),
                      lty = 2) +
  ggplot2::geom_boxplot(varwidth = TRUE) +
  ggplot2::facet_grid(cols = ggplot2::vars(sample_title_with_subgroup),
                      space = "free_x",
                      scales = "free_x") +
  ggplot2::scale_color_viridis_d() +
  ggplot2::labs(title = "Single cell prediction class ratios",
                x = "Sample",
                y = "Relative confidence = ratio of subgroup:others",
                color = "Predicted\nSubgroup") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom")

relative_confidence_plot_object

ggplot2::ggsave(filename = relative_confidence_plot_filepath,
                plot = relative_confidence_plot_object,
                width = 7.5, height = 5)

```
  
### UMAPs with cluster and subgroup
```{r fig.height = 20, fig.width = 7.5}

cluster_confidence_plot_object <- umap_labels_combined_df |>
  ggplot2::ggplot(ggplot2::aes(x = UMAP_1,
                               y = UMAP_2,
                               color = max_score,
                               alpha = confidence)) +
  ggplot2::geom_point(shape = 16) +
  ggplot2::scale_color_viridis_d() +
  ggplot2::scale_alpha_continuous(range = c(0, 1)) +
  ggplot2::facet_wrap(~ sample_title_with_subgroup, ncol = 1,
                      scales = "free") +
  ggplot2::labs(title = "Single cell confidence scores, UMAP",
                x = "UMAP 1",
                y = "UMAP 2",
                alpha = "Confidence =\nmax/total",
                color = "Predicted\nSubgroup") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom",
                 legend.direction = "vertical")

cluster_plot_object <- umap_labels_combined_df |>
  ggplot2::ggplot(ggplot2::aes(x = UMAP_1,
                               y = UMAP_2,
                               color = cluster)) +
  ggplot2::geom_point(shape = 16) +
  ggplot2::facet_wrap(~ sample_title_with_subgroup, ncol = 1,
                      scales = "free") +
  ggplot2::labs(title = "Single cell clusters, UMAP",
                x = "UMAP 1",
                y = "UMAP 2",
                color = "Predicted\nClusters") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom",
                 legend.direction = "vertical")

combined_umap_plot_object <- patchwork::wrap_plots(cluster_plot_object, cluster_confidence_plot, nrow = 1)

combined_umap_plot_object

ggplot2::ggsave(filename = combined_umap_plot_filepath,
                plot = combined_umap_plot_object,
                height = 20,
                width = 7.5)

```

# Session info
```{r session_info}
sessionInfo()
```