---
title: "Predicting tumor subgroup in pseudobulk and single cells"
author: "Steven Foltz"
date: "2023"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    df_print: paged
---

This notebook uses a medulloblastoma subgroup prediction model to predict the subgroup of pseudobulk samples and individual cells from single-cell RNA-seq experiments.
Pseudobulk samples are generated from single-cell experiments by averaging gene expression across all cells on a per gene level.
Prediction models were trained using bulk gene expression data (bulk array and RNA-seq) in `analysis_notebooks/baseline_models.Rmd`.
Here, we use the trained kTSP and RF models to predict single cell subgroups from scRNA-seq data generated by [Hovestadt, et. al](https://www.nature.com/articles/s41586-019-1434-6).


# Setup
```{r setup}

# Load libraries 
source(here::here("utils/color_schemes.R"))
source(here::here("utils/modeling.R"))
source(here::here("utils/single-cell.R"))
library(patchwork)

# Directories
models_dir <- here::here("models")
plots_dir <- here::here("plots")
plots_data_dir <- file.path(plots_dir, "data")
processed_data_dir <- here::here("processed_data")
pseudobulk_sce_dir <- file.path(processed_data_dir, "pseudobulk_sce")

# Input files
pseudobulk_genex_filepath <- file.path(processed_data_dir, "pseudobulk_genex.tsv")
pseudobulk_metadata_filepath <- file.path(processed_data_dir, "pseudobulk_metadata.tsv")

# baseline models are generated prior to this in the notebook analysis_notebooks/baseline_models.Rmd
# there is one baseline model designated as the official model that is used to test single cells here
# "baseline" means that we used all genes and all available training/test data in modeling
baseline_models_filepath <- file.path(models_dir, "baseline.rds")

# Output files
single_cell_plot_data_filepath <- file.path(plots_data_dir, "single_cell_test_predictions.tsv")
pseudobulk_plot_data_filepath <- file.path(plots_data_dir, "pseudobulk_test_predictions.tsv")
pseudobulk_subgroups_plot_filepath <- file.path(plots_dir, "single_cell_pseudobulk_subgroups.pdf")
percentage_subgroups_plot_filepath <- file.path(plots_dir, "single_cell_percentage_subgroups.pdf")
confidence_umap_plot_filepath <- file.path(plots_dir, "single_cell_confidence_umap.pdf")
cluster_umap_plot_filepath <- file.path(plots_dir, "single_cell_cluster_umap.pdf")
combined_percentage_plot_filepath <- file.path(plots_dir, "single_cell_combined_percentage.pdf")
combined_umap_plot_filepath <- file.path(plots_dir, "single_cell_combined_umap.pdf")

# Read in data
pseudobulk_genex_df <- readr::read_tsv(pseudobulk_genex_filepath,
                                       show_col_types = FALSE) |>
  tibble::column_to_rownames(var = "gene")
pseudobulk_metadata_df <- readr::read_tsv(pseudobulk_metadata_filepath,
                                          show_col_types = FALSE) |>
  dplyr::mutate(sample_accession = title)

baseline_models <- readr::read_rds(baseline_models_filepath)
official_model <- which(purrr::map_lgl(baseline_models, \(x) x[["official_model"]]))

classifier_list <- list(ktsp = baseline_models[[official_model]]$ktsp_weighted$classifier,
                        rf = baseline_models[[official_model]]$rf_weighted$classifier)

model_types <- c("ktsp", "rf")
mb_subgroups <- c("G3", "G4", "SHH", "WNT")

```

# Predict the subgroup of pseudobulk samples

```{r message = FALSE}

pseudobulk_test_list <- list("ktsp" = test_ktsp(genex_df_test = pseudobulk_genex_df,
                                                metadata_df_test = pseudobulk_metadata_df,
                                                classifier = classifier_list[["ktsp"]],
                                                labels = mb_subgroups),
                             "rf" = test_rf(genex_df_test = pseudobulk_genex_df,
                                            metadata_df_test = pseudobulk_metadata_df,
                                            classifier = classifier_list[["rf"]]))

```

# Plot pseudobulk subgroup predictions

### Prep plotting data

```{r}

pseudobulk_plot_df <- purrr::map(model_types,
                                 \(model_type) pseudobulk_test_list[[model_type]]$model_output |>
                                   as.data.frame() |>
                                   tibble::rownames_to_column(var = "sample_accession") |>
                                   tibble::as_tibble() |>
                                   tidyr::pivot_longer(cols = all_of(mb_subgroups),
                                                       names_to = "predicted_subgroup") |>
                                   dplyr::select(sample_accession,
                                                 predicted_subgroup,
                                                 value) |>
                                   dplyr::mutate(model_type = model_type)) |>
  dplyr::bind_rows() |>
  dplyr::left_join(pseudobulk_metadata_df |>
              dplyr::select(sample_accession,
                            subgroup),
              by = "sample_accession") |>
  dplyr::mutate(model_type = dplyr::case_match(model_type,
                                               "ktsp" ~ "kTSP (w)",
                                               "rf" ~ "RF (w)"))

readr::write_tsv(x = pseudobulk_plot_df,
                 file = pseudobulk_plot_data_filepath)

```

# Plot pseudobulk subgroup predictions

```{r}

pseudobulk_subgroups_plot_object <- pseudobulk_plot_df |>
  dplyr::filter(model_type == "kTSP (w)") |>
  ggplot2::ggplot(ggplot2::aes(y = sample_accession,
                               x = value,
                               fill = predicted_subgroup)) +
  ggplot2::geom_col(position = "fill") +
  ggplot2::scale_x_continuous(labels = scales::percent) +
  ggplot2::facet_grid(rows = ggplot2::vars(subgroup),
                      space = "free_y",
                      scales = "free_y") +
  ggplot2::scale_fill_manual(values = subgroup_colors) +
  ggplot2::labs(x = "Pseudobulk Score Percentage",
                y = NULL,
                fill = "Predicted Subgroup") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom",
                 legend.direction = "horizontal")

ggplot2::ggsave(filename = pseudobulk_subgroups_plot_filepath,
                plot = pseudobulk_subgroups_plot_object,
                width = 7.5,
                height = 6)

pseudobulk_subgroups_plot_object

```


# Predict the subgroup of single cells

```{r message = FALSE}

model_test_list <- purrr::map(model_types,
                              \(model_type) purrr::map(pseudobulk_metadata_df$title,
                                                       \(x) test_single_cells(sample_acc = x,
                                                                              sce_filepath = file.path(pseudobulk_sce_dir,
                                                                                                       stringr::str_c(x, "_sce.rds")),
                                                                              metadata_df = pseudobulk_metadata_df,
                                                                              labels = mb_subgroups,
                                                                              classifier = classifier_list[[model_type]],
                                                                              model_type = model_type,
                                                                              study = "GSE119926",
                                                                              platform = "scRNA-seq")) |>
                                purrr::set_names(pseudobulk_metadata_df$title)) |>
  purrr::set_names(model_types)

```

# Plot single cell subgroup predictions

### Prep plotting data

```{r}
  
return_umap_cluster <- function(sample_title) {
  
  sce_filepath <- file.path(pseudobulk_sce_dir,
                            stringr::str_c(sample_title, "_sce.rds"))
  
  sce_object <- readr::read_rds(sce_filepath)
  
  umap_df <- data.frame(sample = sample_title,
                        SingleCellExperiment::reducedDim(sce_object, "UMAP"),
                        cluster = SummarizedExperiment::colData(sce_object)$louvain_10) |>
    dplyr::mutate(cell_index = dplyr::row_number())
  
  names(umap_df) <- c("sample_accession", "UMAP_1", "UMAP_2", "cluster", "cell_index")
  row.names(umap_df) <- NULL
  
  return(umap_df)
  
}

umap_list <- purrr::map(pseudobulk_metadata_df$title,
                        \(sample_title) return_umap_cluster(sample_title)) |>
  dplyr::bind_rows()

single_cell_plot_df <- purrr::map(model_types,
                                  \(model_type) purrr::map(model_test_list[[model_type]],
                                                           \(x) dplyr::bind_cols(x$predicted_labels_df,
                                                                                 tibble::as_tibble(x$model_output)) |>
                                                             tidyr::separate_wider_delim(cols = sample_accession,
                                                                                         delim = "_",
                                                                                         names = c("sample_accession",
                                                                                                   "cell_index"))) |>
                                    dplyr::bind_rows()) |>
  purrr::set_names(model_types) |>
  dplyr::bind_rows(.id = "model_type") |>
  dplyr::rowwise() |>
  dplyr::mutate(max = max(G3, G4, SHH, WNT),
                total = sum(G3, G4, SHH, WNT),
                confidence = max/total,
                cell_index = as.numeric(cell_index)) |>
  dplyr::ungroup() |>
  dplyr::left_join(umap_list,
                   by = c("sample_accession", "cell_index")) |>
  dplyr::left_join(pseudobulk_metadata_df |>
                     dplyr::select(sample_accession,
                                   study,
                                   subgroup,
                                   is_PDX),
                   by = "sample_accession")

readr::write_tsv(x = single_cell_plot_df,
                 file = single_cell_plot_data_filepath)

```

### Plot percentage of subgroups

```{r}

percentage_subgroups_plot_object <- single_cell_plot_df |>
  dplyr::filter(model_type == "ktsp") |>
  dplyr::count(sample_accession, subgroup, predicted_labels) |>
  ggplot2::ggplot(ggplot2::aes(y = sample_accession,
                               x = n,
                               fill = predicted_labels)) +
  ggplot2::geom_col(position = ggplot2::position_fill()) +
  ggplot2::scale_x_continuous(labels = scales::percent) +
  ggplot2::facet_grid(rows = ggplot2::vars(subgroup),
                      space = "free_y",
                      scales = "free_y") +
  ggplot2::scale_fill_manual(values = subgroup_colors) +
  ggplot2::labs(y = NULL,
                x = "Single-cell Percentage",
                fill = "Predicted Subgroup") +
  ggplot2::theme_bw() +
  ggplot2::theme(legend.position = "bottom")

ggplot2::ggsave(filename = percentage_subgroups_plot_filepath,
                plot = percentage_subgroups_plot_object,
                height = 10,
                width = 7.5)

percentage_subgroups_plot_object

```

- Most samples show the same tumor subgroup predicted for most of their cells
- Three samples ("BCH825", "Med2312FH", "SJ625") show a mixture of predicted subgroups (G3/G4)

# Plot confidence of subgroup predictions

### Prep plotting data

Three samples in particular are of interest given the mixture of subgroups predicted ("BCH825", "Med2312FH", "SJ625").
Four other samples can serve as representative samples from each subgroup ("BCH1205", "BCH1031", "MUV41", "BCH807").

```{r}

mixed_samples <- sort(c("BCH825", "Med2312FH", "SJ625"))

mixed_samples_with_subgroup <- pseudobulk_metadata_df |>
  dplyr::filter(sample_accession %in% mixed_samples) |>
  dplyr::mutate(sample_with_subgroup = stringr::str_c(title, " (", subgroup, ")")) |>
  dplyr::select(sample_accession, subgroup, sample_with_subgroup)

```

### Plot UMAPs with subgroup prediction

```{r}

confidence_umap_plot_object <- single_cell_plot_df |>
  dplyr::filter(model_type == "ktsp",
                sample_accession %in% mixed_samples) |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  ggplot2::ggplot(ggplot2::aes(x = UMAP_1,
                               y = UMAP_2,
                               color = predicted_labels)) +
  ggplot2::geom_point(shape = 16,
                      alpha = 0.5) +
  ggplot2::scale_color_manual(values = subgroup_colors) +
  ggplot2::scale_alpha_continuous(range = c(0, 1)) +
  ggplot2::facet_wrap(~ sample_with_subgroup,
                      nrow = 1,
                      scales = "free") +
  ggplot2::labs(title = "Single cell confidence scores, UMAP",
                x = "UMAP 1",
                y = "UMAP 2",
                color = "Predicted\nSubgroup") +
  ggplot2::theme_bw() +
  ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(alpha = 1)))

ggplot2::ggsave(filename = confidence_umap_plot_filepath,
                plot = confidence_umap_plot_object,
                width = 7.5,
                height = 6)

confidence_umap_plot_object
```

### Plot UMAPs with subgroup prediction

```{r}

cluster_umap_plot_object <- single_cell_plot_df |>
  dplyr::filter(model_type == "ktsp",
                sample_accession %in% mixed_samples) |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  ggplot2::ggplot(ggplot2::aes(x = UMAP_1,
                               y = UMAP_2,
                               color = cluster)) +
  ggplot2::geom_point(shape = 16,
                      alpha = 0.5) +
  ggplot2::facet_wrap(~ sample_with_subgroup,
                      nrow = 1,
                      scales = "free") +
  ggplot2::scale_color_brewer(palette = "Set1") +
  ggplot2::labs(title = "Single cell clusters, UMAP",
                x = "UMAP 1",
                y = "UMAP 2",
                color = "Cluster") +
  ggplot2::theme_bw() +
  ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(alpha = 1)))

ggplot2::ggsave(filename = cluster_umap_plot_filepath,
                plot = cluster_umap_plot_object,
                width = 7.5,
                height = 6)

cluster_umap_plot_object
```

### Plot confidence of individual cell predictions

```{r}

typical_confidence <- single_cell_plot_df |>
  dplyr::filter(model_type == "ktsp",
                !is.na(confidence)) |>
  dplyr::group_by(sample_accession,
                  subgroup,
                  predicted_labels) |>
  dplyr::summarize(median_confidence = median(confidence),
                   .groups = "drop") |>
  tidyr::pivot_wider(names_from = predicted_labels,
                     values_from = median_confidence) |>
  dplyr::rowwise() |>
  dplyr::mutate(subgroup_conf = c(G3, G4, SHH, WNT)[which(mb_subgroups == subgroup)]) |>
  dplyr::pull(subgroup_conf) |>
  median()

cell_counts_df <- single_cell_plot_df |>
  dplyr::filter(model_type == "ktsp",
                sample_accession %in% mixed_samples,
                !is.na(confidence)) |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  dplyr::count(sample_with_subgroup,
               predicted_labels)

confidence_plot_object <- single_cell_plot_df |>
  dplyr::filter(model_type == "ktsp",
                sample_accession %in% mixed_samples,
                !is.na(confidence)) |>
  dplyr::left_join(mixed_samples_with_subgroup,
                   by = "sample_accession") |>
  ggplot2::ggplot(ggplot2::aes(x = predicted_labels,
                               y = confidence,
                               color = predicted_labels)) +
  ggplot2::geom_hline(yintercept = typical_confidence,
                      lty = 2) +
  ggplot2::geom_violin(fill = NA,
                       draw_quantiles = 0.5,
                       position = ggplot2::position_dodge(width = 0.75), scale = "width",
                       show.legend = FALSE) +
  ggplot2::geom_point(shape = 16,
                      alpha = 0.5,
                      position = ggplot2::position_jitterdodge(jitter.height = 0,
                                                               jitter.width = 0.25),
                      show.legend = FALSE) +
  ggplot2::geom_text(data = cell_counts_df,
                     ggplot2::aes(label = n),
                     y = 0,
                     vjust = 0,
                     show.legend = FALSE) +
  ggplot2::facet_wrap(~ sample_with_subgroup,
                      nrow = 1,
                      scales = "free") +
  ggplot2::scale_color_manual(values = subgroup_colors) +
  ggplot2::scale_y_continuous(limits = c(0, NA)) +
  ggplot2::labs(title = "Single cell prediction confidence",
                x = "Predicted Subgroup",
                y = "Confidence",
                color = "Predicted Subgroup") +
  ggplot2::theme_bw() +
  ggplot2::guides(colour = ggplot2::guide_legend(override.aes = list(alpha = 1)))

ggplot2::ggsave(filename = confidence_umap_plot_filepath,
                plot = confidence_plot_object,
                width = 7.5, height = 5)

confidence_plot_object

```

- Takeaway: Cell subgroup predictions are more confident when subgroup predictions are consistent across entire sample.
- The relative confidence levels of each cell's prediction show that predictions in samples with one dominant subgroup are more confident than in samples with mixed subgroup predictions (BCH825, Med2312FH, SJ625).
- While some cells from mixed subgroup samples show high confidence as G3 or G4, most show that cells receive signals from both G3 and G4 rules, meaning they have an intermediate phenotype, rather than being a mixture of true G3 and true G4 cells.

# Combined plots

```{r}
combined_percentage_plot_object <- pseudobulk_subgroups_plot_object + percentage_subgroups_plot_object +
  patchwork::plot_layout(guides = "collect") +
  patchwork::plot_annotation(tag_levels = 'a') &
  ggplot2::theme(legend.position = "bottom")

ggplot2::ggsave(filename = combined_percentage_plot_filepath,
                plot = combined_percentage_plot_object,
                width = 7.5,
                height = 6)

combined_percentage_plot_object

```

```{r}

combined_umap_plot_object <- confidence_umap_plot_object / cluster_umap_plot_object / confidence_plot_object +
  patchwork::plot_annotation(tag_levels = 'a')
  
ggplot2::ggsave(filename = combined_umap_plot_filepath,
                plot = combined_umap_plot_object,
                width = 7.5,
                height = 6)

combined_umap_plot_object

```

# Session info
```{r session_info}
sessionInfo()
```
